### ISSUE TITLE
Gradle 6.8 incorrectly swaps binary dependency to project dependency

### ISSUE THREAD

**Issue by liutikas** [gradle/gradle#15781]
When upgrading from Gradle 6.7.1 to 6.8 we found that our project started failing with
```
aurimas@aurimas:/ssd/ssd1/temp/testgradle68$ ./gradlew arch:core:core-common:jar
Watching the file system is not supported on this operating system.
Configuration on demand is an incubating feature.

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':arch:core:core-common:compileJava'.
> Could not resolve all task dependencies for configuration ':arch:core:core-common:compileClasspath'.
   > Could not resolve androidx.annotation:annotation:1.1.0.
     Required by:
         project :arch:core:core-common
      > No matching configuration of project :annotation:annotation was found. The consumer was configured to find an API of a library compatible with Java 7, preferably in the form of class files, and its dependencies declared externally but:
          - None of the consumable configurations have attributes.
```

There were no other changes to our build outside of Gradle upgrade.

This project uses the following dependency
```
dependencies {
    api("androidx.annotation:annotation:1.1.0")
}
```

### Expected Behavior
Gradle uses binary dependency for this project

### Current Behavior
Gradle swaps this project dependency to be `project :annotation:annotation`

### Context
There seems to be several factors at play that are required to reproduce this failure
- Use of `includeBuild`
- Rename of project in `settings.gradle`
- Use of `org.gradle.configureondemand=true`

It works just fine in Gradle 6.7.1 and it fails with 6.8

### Steps to Reproduce 
Attaching a sample project
[testgradle68.zip](https://github.com/gradle/gradle/files/5810783/testgradle68.zip)

To repro simply run:
```
./gradlew arch:core:core-common:jar
```

If you downgrade to 6.7.1 it will work again.

### Your Environment
Unable to share full details due to corp policy.

**Comment by ljacomet** [comment]
This is an unexpected consequence of improvements done around composite builds.

What happens is that the default coordinates for project `:annotation:annotation` are computed as follows:
* `group` takes the value `<rootProjectName>.<parentProjectName>` which results in `androidx.annotation`
* `name` is the project name, `annotation`
This results in an identifier `androidx.annotation:annotation` which matches the binary dependency coordinates.

The above is the historical default and is not new.

What is new is that projects of the main build are now registered as substitutions whenever you have a composite build. This is what causes the binary dependency to be replaced by a project reference.

Then the dependency fails to resolve because Gradle does not know `:annotation:annotation` participates in the build and since configuration-on-demand is active, it ends up not configured, which means it does not have any configurations defined.

The team is discussing reverting this change altogether or hiding it behind a feature preview flag for now.

**Comment by jjohannes** [comment]
Responsible change: #14638

**Comment by jjohannes** [comment]
The substitutions are now only registered if the main build is explicitly included. That should fix this issue in 6.8.1 - see #15817.
